Mage/src/main/java/mage/game/events/GameEvent.java

```diff
@@ -360,8 +360,19 @@ playerId    player who try to targeting (can be different from source ability's
          combine all TAPPED events occuring at the same time in a single event
          */
         TAPPED_BATCH,
-
-        UNTAP, UNTAPPED,
+        UNTAP,
+        /* UNTAPPED,
+         targetId    untapped permanent
+         sourceId    not used for this event // TODO: add source for untap?
+         playerId    controller of permanent // TODO: replace by source controller of untap? need to check every usage if so.
```

Comment:
not sure of that TODO, could remove it. source would make sense if needed to extract info there.

Reply:
For info

1.

It can be refactor someday to add `source` info to `untap` and other calls (see 35f5a8257b14212b72b9ecb88b624f998a914fe9 for example). It's a standard code style in xmage -- send source and game params to game engine methods like:
* `doAction(Ability source, Game game, other params)`.

Also a `source` param logic:
* non nullable source for all effects;
* nullable source for state base actions (SBA);

![shot_231029_034732](https://github.com/magefree/mage/assets/8344157/af31d311-22db-4140-b448-0a9888b7b106)

2.

> TODO: replace by source controller of untap? need to check every usage if so.

Be careful, current `playerId` contains permanent's controller, not source's controller:
![shot_231029_040044](https://github.com/magefree/mage/assets/8344157/96843c2e-4bac-452e-8395-55f3f12dbb9f)

But I check untap event usage and it's safe to replace a logic -- current effects uses only event's targetId param:
![shot_231029_040604](https://github.com/magefree/mage/assets/8344157/9b9aebc6-c28e-44e2-8010-54e543194b6b)

Reply:
BTW `event.getSourceId() == null` logic can be used instead untap step flag too.

Reply:
Good points. I'll make a followup issue about this. For now I'd rather focus on implementing the new set that doing refactor.